STYLEGUIDE.md

```diff
@@ -0,0 +1,740 @@
+The verb should fit the return type and value.
+An exception to this can be methods, that operate on an instance directly, e. g. `instance.toString()`, `instance.asObject()`.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use functions without verb */
+function toFoo(): Foo {
+  ...
+}
+
+/* Avoid verbs that imply a different return type or behavior */
+function isFoo(): void {
+  ...
+  throw Error(...);
+}
+
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use functions with verbs */
+function transformToFoo(): Foo {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function isFoo(): boolean {
+  ...
+}
+
+/* Use functions with verbs that fit the return type */
+function validateFoo(): void {
+  ...
+  throw Error(...);
+}
+
+/* Exception, you can omit verbs in methods, that transform the instance that they are invoked on */
+class FooBar {
+  toString(): string {
+    ...
+  }
+
+  asObject(): Record<string, any> {
+    ...
+  }
+}
+```
+
+## Strings
+
+### Use single quotes ✓
+
+Use single quotes, unless not possible otherwise, e. g. when your string contains single quotes.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use double quotes */
+const foo = "foo";
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use single quotes */
+const foo = 'foo';
+
+/* Use double quotes if the quoted string contains single quotes. */
+const foo = "'foo'";
+
+/* Use template string if the quoted string contains double and single quotes. */
+const foo = `They know when to use 'single quotes' and "double quotes"`;
+### String concatenation
+When possible use template literals, this makes the code more readable in most cases.
+Feel free to concatenate strings with `+` in case it makes to code better readable.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't concatenate strings with `+` */
+const foo = 'foo ' + bar + ' bar';
+
+/* Gray area: If template literals become too complex, don't use them. */
+const foo = `foo ${fn(`${bar}bar`)}`;
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Concatenate strings using template literals */
+const foo = `foo ${bar} bar`;
+
+/* Gray area: If template literals become too complex, feel free to concatenate with +. */
+const foo = `foo ${fn(bar + 'bar')}`;
+```
+
+## Arrays
+
+### Use `.forEach` rather than `for`
+Use the functional language concepts of JavaScript in favor of the imperative ones, when possible.
+Use the correct function for your use case.
+In rare cases, when performance is of the essence, it makes sense to resort to the imperative `for` instead.
+
+❌ Examples of **incorrect** code:
+```ts
+/* Don't use for ... of if you can use .forEach */
+for (const item of array) {
+  ...
+}
+
+/* Don't use .forEach if you can use .map */
+const result = [];
+array.forEach(item => {
+  result.push(item.property);
+});
+```
+
+✅ Examples of **correct** code:
+```ts
+/* Use .forEach if possible */
+array.forEach(item => {
+  ...
+});
+
+/* Use .map, if appropriate */
```

Comment:
```suggestion
});

/* Exceptions: when conditional break is used */
for (let i=0;;i++) {
  if (i>10)
    break;
}

/* Use .map, if appropriate */
```

Reply:
@jjtang1985 I was thinking whether we can find a more plausible example, but I can't think of any case, where a for loop would be better than one of the array functions.

I would rather write your example from above as:
```
array.slice(0, 10).forEach(...)
```

Reply:
you are right. Let's update when I have more appropriate example.